tests/ui/lint/non_local_definitions.rs

```diff
@@ -0,0 +1,117 @@
+// check-pass
+// edition:2021
+
+use std::fmt::{Debug, Display};
+
+struct Test;
+
+impl Debug for Test {
+    fn fmt(&self, _f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        todo!()
+    }
+}
+
+mod do_not_lint_mod {
+    pub trait Tait {}
+
+    impl super::Test {
+        fn hugo() {}
+    }
+
+    impl Tait for super::Test {}
+}
+
+const _: () = {
+    trait Uto {}
+
+    impl Uto for Test {}
+};
+
+trait Uto1 {}
+enum Enum {
+    Discr = {
+        impl Uto1 for Test {}
+        //~^ WARN non local definition
+
+        1
+    }
+}
+
+trait Uto2 {}
+static A: u32 = {
+    impl Uto2 for Test {}
+    //~^ WARN non local definition
+
+    1
+};
+
+trait Uto3 {}
+const B: u32 = {
+    impl Uto3 for Test {}
+    //~^ WARN non local definition
+
+    #[macro_export]
+    macro_rules! m0 { () => { } };
+    //~^ WARN non local definition
+
+    trait Uto4 {}
+    impl Uto4 for Test {}
+
+    1
+};
+
+trait Uto5 {}
+fn main() {
+    #[macro_export]
+    macro_rules! m { () => { } };
+    //~^ WARN non local definition
+
+    impl Test {
+    //~^ WARN non local definition
+        fn foo() {}
+    }
+
+    impl Display for Test {
+    //~^ WARN non local definition
+        fn fmt(&self, _f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+            todo!()
+        }
+    }
+
+    let _a = || {
+        impl Uto5 for Test {}
+        //~^ WARN non local definition
+
+        1
+    };
+
+    struct InsideMain;
+
+    impl Debug for InsideMain {
+        fn fmt(&self, _f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+            todo!()
+        }
+    }
+
+    impl InsideMain {
+        fn foo() {}
+    }
+
+    fn inside_inside() {
+        impl Display for InsideMain {
+        //~^ WARN non local definition
+            fn fmt(&self, _f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+                todo!()
+            }
+        }
+
+        impl InsideMain {
+        //~^ WARN non local definition
+            fn bar() {
+                #[macro_export]
+                macro_rules! m2 { () => { } };
+                //~^ WARN non local definition
+            }
+        }
+    }
```

Comment:
It would be cool to also have a test for inline const and array lengths.

Reply:
Done.